src/kontrol/foundry.py

```diff
@@ -722,6 +725,18 @@ def free_proof_version(
         latest_version = self.latest_proof_version(test)
         return latest_version + 1 if latest_version is not None else 0
 
+    def remove_old_proofs(self, force_remove: bool = False) -> bool:
+        if force_remove or any(
+                not method.contract_up_to_date(Path(method.contract_digest))
+                for contract in self.contracts.values()
+                for method in contract.methods
+            ):
+                shutil.rmtree(self.proofs_dir.absolute())
+                return True
+        else:
+            return False
```

Comment:
Wouldn't it make more sense for this function to delete only the outdated proof dirs?

Also, I'd say it is redundant to check the `contract_up_to_date` on each method of a contract. You could either:
1. move the `contract_up_to_date` and `contract_digest` methods in the `Contract` class instead of the `Method` and then just do
```python
not contract.contract_up_to_date(Path(contract_digest))
for contract in self.contracts.values()
```
2. (might be easier to implement now) Check `contract_up_to_date` only for the first function of a contract.
```py
  not contract.methods[0].contract_up_to_date(Path(contract.methods[0].contract_digest))
for contract in self.contracts.values()
```

Reply:
Thank you for the recommendation!

Isn't the digest based on all functions of the contract? Namely, if a function changes or a new one is added, the digest changes? Asking to know how exactly would targeting the individual outdated proof dirs work!

Reply:
A problem with option 2 might be that if a contract doesn't have any methods it'll throw an error when trying to access `contract.methods[0]` I think

Reply:
> Isn't the digest based on all functions of the contract? Namely, if a function changes or a new one is added, the digest changes?

That's my understanding as well. In addition, there is a `method` digest, and a `contract` digest. AFAIK, the `contract` digest is the same for any method of that contract.

edit:
here's a digest example
```json
{
    "methods": {
        "test%CounterTest.test_Increment()": {
            "method": "4c32041fa5a7c64776f36dab5ae8cc61588efc3e3fb09b3c95ed0917587687d6",
            "contract": "29c961bb216d9a5f16961d6a224fd6cb1feeb18b36e29e9a09c2733b0590a655"
        },
        "test%CounterTest.setUp()": {
            "method": "2c1737186c0be7909e9db344314dabfa10f5f492345ad17e87375ed13c1bcaa6",
            "contract": "29c961bb216d9a5f16961d6a224fd6cb1feeb18b36e29e9a09c2733b0590a655"
        },
        "test%CounterTest.testFuzz_SetNumber(uint256)": {
            "method": "323d5de4c923be8a393316579d3833bd4f00d4d81a5a85b3030a54efd2222e87",
            "contract": "29c961bb216d9a5f16961d6a224fd6cb1feeb18b36e29e9a09c2733b0590a655"
        }
    },
    "kompilation": "089cb7a3c95b4643e44147c41139b379b7a69797b3a8cb3d6a9e4acc41fb3bef",
    "kontrol": "0.1.32",
    "foundry": "a7794d1db53643ecf7e89e8be01ca0fcad6ca2cf381e68f4fdbe70d1be6934f9"
}
```
For any method of `CounterDigest` the `contract` hash is the same. So I don't see a point to loop through each function of the contract.

Reply:
According to [the docs](https://docs.python.org/3/library/functions.html#any), `any` (and `all`, for that matter) evaluate lazily. So we wouldn't be computing all of the `contract_up_to_date` and then seeing if any of them are not up to date. It'd stop at the first positive.
In light of this, my previous comment, and that factoring out the contract digest logic down to the `Contract` class seems to deserve a PR of its own, I'd suggest leaving the code as is. What do you think?

Reply:
Yes, sounds good. We should open an issue for this!

Reply:
Yep!! On it!